src/shims/io.rs

```diff
@@ -123,38 +144,71 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let this = self.eval_context_mut();
 
         if !this.machine.communicate {
-            throw_unsup_format!("`open` not available when isolation is enabled")
+            throw_unsup_format!("`read` not available when isolation is enabled")
         }
 
         let tcx = &{ this.tcx.tcx };
 
-        let fd = this.read_scalar(fd_op)?.to_i32()?;
-        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
         let count = this.read_scalar(count_op)?.to_usize(&*this.tcx)?;
+        // Reading zero bytes should not change `buf`
+        if count == 0 {
+            return Ok(0);
+        }
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let buf_scalar = this.read_scalar(buf_op)?.not_undef()?;
 
         // Remove the file handle to avoid borrowing issues
-        this.remove_handle_and(
-            fd,
-            |mut handle, this| {
-                let bytes = handle
-                    .file
-                    .read(this.memory_mut().get_mut(buf.alloc_id)?.get_bytes_mut(
-                        tcx,
-                        buf,
-                        Size::from_bytes(count),
-                    )?)
-                    .map(|bytes| bytes as i64);
-                // Reinsert the file handle
-                this.machine.file_handler.handles.insert(fd, handle);
-                this.consume_result(bytes)
-            },
-        )
+        this.remove_handle_and(fd, |mut handle, this| {
+            // Don't use `?` to avoid returning before reinserting the handle
```

Comment:
Why do we even remove and re-insert, instead of working with a borrowed handle?

Reply:
because of borrowing issues, we'd have to mutably borrow the handle, and then mutably borrow the bytes inside memory.

Reply:
Could this be fixed if we made the `memory` field public, instead of exposing it just via getters?

Reply:
My naive opinion is that it would work because `memory` and `machine` are different fields of `ecx`. So you can get a mutable reference to file handle from `machine` and the mutable reference to the bytes from `memory`.

Reply:
Great. I'll make a PR.